plantcv/plantcv/transform/warp.py

```diff
@@ -61,51 +85,99 @@ def warp(img, refimg, pts, refpts, method='default'):
+                               markerSize=params.marker_size * res_ratio_i,
+                               thickness=params.line_thickness * res_ratio_i)
+
+        # rgb image for colored markers on refimg
+        refimg_marked = refimg_.copy()
         if len(shape_ref) == 2:
-            refimg2 = cv2.merge((refimg2, refimg2, refimg2))
+            refimg_marked = cv2.cvtColor(refimg_marked, cv2.COLOR_GRAY2RGB)
+
         for i, pt in enumerate(refpts):
-            cv2.drawMarker(refimg2,
-                           pt,
-                           color=colors[i],
-                           markerType=cv2.MARKER_CROSS,
-                           markerSize=params.marker_size*res_ratio_r,
-                           thickness=params.line_thickness*res_ratio_r)
+            if status[i][0] == 1:
+                cv2.drawMarker(refimg_marked, pt, color=colors[i], markerType=cv2.MARKER_CROSS,
+                               markerSize=params.marker_size * res_ratio_r,
+                               thickness=params.line_thickness * res_ratio_r)
+            else:
+                cv2.drawMarker(refimg_marked, pt, color=colors[i], markerType=cv2.MARKER_TRIANGLE_UP,
+                               markerSize=params.marker_size * res_ratio_r,
+                               thickness=params.line_thickness * res_ratio_r)
 
         debug_mode = params.debug
         params.debug = None
-        img_blend = overlay_two_imgs(warped_img, refimg)
+
+        # make sure the input image for "overlay_two_imgs" is of dtype "uint8" such that it would be acceptable for overlay_two_imgs (cv2.cvtColor)
+        img_blend = overlay_two_imgs(_preprocess_img_dtype(warped_img), refimg_)
         params.debug = debug_mode
 
-        if params.debug == 'plot':
-            plot_image(img2)
-            plot_image(refimg2)
-            plot_image(img_blend)
-        if params.debug == 'print':
-            print_image(img2, os.path.join(params.debug_outdir, str(params.device) + "_img-to-warp.png"))
-            print_image(refimg2, os.path.join(params.debug_outdir, str(params.device) + "_img-ref.png"))
-            print_image(img_blend, os.path.join(params.debug_outdir, str(params.device) + "_warp_overlay.png"))
+        _debug(visual=img_marked, filename=os.path.join(params.debug_outdir, str(params.device) + "_img-to-warp.png"))
+        _debug(visual=refimg_marked, filename=os.path.join(params.debug_outdir, str(params.device) + "_img-ref.png"))
+        _debug(visual=img_blend, filename=os.path.join(params.debug_outdir, str(params.device) + "_warp_overlay.png"))
+
+    return warped_img, mat
+
+
+def warp_align(img, mat, refimg=None):
+    """
+    Warp the input image based on given transformation matrix mat, to align with the refimg
+
+    Inputs:
+    img = image to be warped
+    mat = transformation matrix (size: (3,3))
+    refimg = (option) reference image. default=None
```

Comment:
the output of this function doesn't make much sense if refimg isn't set here. 
![image](https://user-images.githubusercontent.com/7461221/116484518-17764e00-a83e-11eb-8124-1865de1e8691.png)
it seems reasonable to me to require the output image size if warping an image.

Reply:
Right, I do think we need to change this somehow. The size of the output image is determined by the `refimg`, so if one is not provided then the transformed image is output at its original size. While that works it does not produce an image that is likely usable in many cases (assuming that typically `img` and `refimg` are different sizes).

One option is to require `refimg`. Another is to require the output shape and not have a `refimg`, which is only used to get the shape and plot an overlay (not needed because fit was already determined by calculating the matrix earlier?).

Reply:
it comes down to which one is more annoying:

shape is required and you have refimg:
`refimg.shape`
or
refimg is required and you don't have refimg:
`np.zeros_like((x,y))`

I would assume the user has refimg, for example an image from another camera you are matching.
i'm flexible though.

Reply:
True, plus if desired someone could do `img=img, refimg=img`

Reply:
This `warp_align` function is there for the cases where the transformation matrix is available, such that we can warp the `img` with or without `ref_img`. The `ref_img` is only used for visualization - the blended image. The returned variable is the `warped_img`

Reply:
I was thinking the function of `warp_align` was more that you don't need to recalculate `mat` for each pair of images, you would just need to create it once with a pair of training images. But then on each use, you would still have the pair of `img` and `refimg` (e.g. thermal + RGB, RGB + NIR). 

When you create `mat` based on images of different sizes, if you run `warp_align` without `refimg` you will get an image of size `img` instead of `refimg`, so then it likely won't work for what you want to do with it downstream.

But in the event one really doesn't have a `refimg`, either `img` can be input as both `img` and `refimg` or a blank image using `np.zeros` can be input as `refimg`.

Reply:
That's correct. For most of the prospective user cases, both `img` and `refimg` are expected inputs. 
Oh, I see the point, 
`warped_img = cv2.warpPerspective(src=img, M=mat, dsize=(cols_ref, rows_ref))`

The size of the warped image depends on `mat` as well as provided `(cols_ref, rows_ref))`